[Quantization] Add metadata porting for nodes added by quantization

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107107
* #107900
* #107899
* #107106
* #107105

Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

kimishpatel committed:
[Quantization] Add metadata porting for nodes added by quantization

Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,215 @@
+__all__ = ["PortNodeMetaForQDQ"]
+
+_METADATA_TO_PORT = [
+    "nn_module_stack",
```
what about source_fn and val?

kimishpatel replied:
source_fn is not valid because it also contains nn.Module and I dont want someone to inadvertently grab nodes belonging to say nn.Linear, and now those nodes also include q/dq nodes.

Val is fair one but then we would have to do that as part of convert workflow which should actually quantize the value and dequantize it. wdyth

jerryzh168 replied:
OK, yeah I think it makes sense to not add source_fn here, and should the same reasoning should apply for nn_module_stack as well? e.g. people might be looking for things in nn.Linear module scope, then it will get q/dq ops, will this be surprising? not too familiar with "stack_trace", maybe that one is fine

I thought "val" is storing shape and dtypes, is that true? yeah we do need to update the dtype I think

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,215 @@
+          - [Q-> [DQ -> Conv -> Q] -> [DQ -> AvgPool -> Q] -> [DQ -> Linear -> Q] -> DQ]
+          - Note first Q and last DQ do not inherit metadata from any nodes
+      - Dynamically quantized patterns:
+        - Input that are dynamically quantized have chose_qparams, quantize and dequantize nodes
```
nit chose_qparams --> choose_qparams

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,215 @@
+
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        for node in graph_module.graph.nodes:
+            annotation = node.meta.get("quantization_annotation", None)
+            if _is_valid_annotation(annotation):
+                input_qspec_map = node.meta["quantization_annotation"].input_qspec_map
+                output_qspec = node.meta["quantization_annotation"].output_qspec
```
do we need to care about annotation in convert step?

kimishpatel replied:
I think the reason annotation is inspected is to that you can distinguish between these uses cases

conv -> add -> linear: quantize with all 3 nodes annotated, you get
q -> dq -> conv -> q -> dq -> add -> q -> dq -> linear -> q -> dq. here quantized op patterns are in brackets
q -> [dq -> conv -> q] -> [dq -> add -> q] -> [dq -> linear -> q] -> dq

While only quantizing conv and linear, i.e. no annotation on add you get
q -> dq -> conv -> q -> dq -> add -> q -> dq -> linear -> q -> dq: quantized op patterns
q -> [dq -> conv -> q] -> dq -> add -> q -> [dq -> linear -> q] -> dq


Both graphs are exactly the same but quantized op patterns are different because in the latter case you dont want add to be quantized. Without inspecting quantization annotation, we cannot know if add was meant to be quantized or not.

jerryzh168 replied:
I see, is the goal here for annotation to control lowering? I feel annotation should not be coupled with lowering, can we just have another api in lowering step for lowering users to control how they want their model to be lowered?

jerryzh168 replied:
like you mentioned before, people can choose to use or not use these tags for lowering, these just expresses what quant flow counts as quantized op, since even with the same tags, people can choose to lower differently

jerryzh168 replied:
another thing I'm thinking of is that I think we should have each step self contained:

1. original model + quantizer
2. annotated model (now the information from quantizer should be 100% transferred to annotation
3. prepared model (now annotation information should be 100% transferred to prepared model, to observer/fake quants)
4. converted model (now prepared model information should be 100% transferred to converted model)

otherwise there will be extra burdens added to serialization I think. for example if in step 4. we expect quantization annotation to be there, then that means the quantization annotation need to be serialized in prepared model as well, and the prepared model is not self contained by itself.

kimishpatel replied:
> I see, is the goal here for annotation to control lowering? I feel annotation should not be coupled with lowering, can we just have another api in lowering step for lowering users to control how they want their model to be lowered?

Annotation is expressing the intent of what is to be quantized. And if this intent is not communicable to lowering stage, it is a problem.

> like you mentioned before, people can choose to use or not use these tags for lowering, these just expresses what quant flow counts as quantized op, since even with the same tags, people can choose to lower differently

Yes, but with porting of tags it makes it much easier to decode user's intent of what was to be quantized and what was not. I can imagine during delegation, the tags are used to figure out what would be run on cpu and what on special backend.

kimishpatel replied:
> otherwise there will be extra burdens added to serialization I think. for example if in step 4. we expect quantization annotation to be there, then that means the quantization annotation need to be serialized in prepared model as well, and the prepared model is not self contained by itself.

That is true. So the question is on serialization and what part of meta will be serialized or not.

Alternatively you have to capture the intent, as expressed by annotations, in the IR itself (if meta is not considrered part of IR). But not that even FX QAT today requires you to prepare the model again and if you chose and different way of doing prepare then there is a loss. So I dont think it is regressing from that behavior anyway. However, I do think preserving annotations somehow through serialization is important. Otherwise intent is lost. Now if that happens via meta serialization or otherwise, is upto what is possible.

jerryzh168 replied:
understood. I think we need to align a bit on the role of quantization_tags I think. right now my understanding is even without quantization_tag, user should be able to lower, since the graph is the ground truth and it should contain all the information they need for lowering. in that case quantization_tag is an optional utility that expresses quant_flows view on these ops. and user can choose not to follow the advice of quantization_tag and lower differently, e.g. they can lower [q -> dq -> conv -> q -> dq] to be a single op.

looks like you were saying it is a mandatory thing that all users should be aware of in order to express the intent?

kimishpatel replied:
> should be aware of in order

I am not saying it is mandatory or not.

There are two possibilities
1. Lower using IR only. Loss of user's intent
2. Lower using IR + something else. QUantization_tag or annotations. Likely no loss of user's intent. quantization_tag allows you to do preserve users' intent, but is not the only way to do that. It just makes it bit more convenient.

jerryzh168 replied:
to clarify, what is the exact definition of user's intent? it seems to make perfect sense in the cases above and for the cases that user just want to lower [dq -> fp32_op -> q] patterns. but what about the pattern I mentioned [q -> dq -> conv -> q -> dq], people might annotate the conv node the exact same way as if they want to lower with [dq -> conv -> q] pattern, but how can they express the lowering intent of [q -> dq -> conv -> q -> dq]?

I feel there is intent of inserting q/dq ops, and intent of lowering. right now it looks like we are trying to express both in the annotation API and I feel not all lowering intent can be expressed with the current design and also feel it's cleaner to separate the inserting q/dq ops and lowering intent for a most general and extensible design.

jerryzh168 replied:
TBH I do feel it's easier UX if annotation is just expressing quantized ops directly, but seems like the reality is there is a separation between inserting q/dq ops and lowering and they are just separate, even though a lot of cases they are coupled

kimishpatel replied:
> to clarify, what is the exact definition of user's intent? it seems to make perfect sense in the cases above and for the cases that user just want to lower [dq -> fp32_op -> q] patterns. but what about the pattern I mentioned [q -> dq -> conv -> q -> dq], people might annotate the conv node the exact same way as if they want to lower with [dq -> conv -> q] pattern, but how can they express the lowering intent of [q -> dq -> conv -> q -> dq]?

In case of "conv -> add -> linear", user doesnt want to quantize add but wants to quantize conv. This is not about lowering intent preservation as much as associating q/dq nodes. In quantizing conv->add->linear where resulting pattern is `q1 -> dq1 -> conv -> q2 -> dq2 -> add -> q3 -> dq3 -> linear -> q4 -> dq4`, dq2 and q3 are not associated with add because add was not be quantized. That is all this is about. SO I am not sure what part is not clear.

jerryzh168 replied:
might be easier to explain in a meeting, let's follow up

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,215 @@
+            if _is_valid_annotation(annotation):
+                input_qspec_map = node.meta["quantization_annotation"].input_qspec_map
+                output_qspec = node.meta["quantization_annotation"].output_qspec
+                for input_node, qspec in input_qspec_map.items():
+                    _port_metadata_for_input_quant_nodes(input_node, node, qspec)
+                _port_metadata_for_output_quant_nodes(node, output_qspec)
```
I'm wondering if it is possible to just look at the graph and port the data, to reduce dependencies, we don't have guarantees that quantization_annotation on nodes will be preserved to this point right now I feel

jerryzh168 replied:
with the current rules, what I can think of is:

1. find choose_qparams -> q -> dq pattern and port metadata
2. port metadata for q nodes
3. port metadata for dq nodes

and there will be some branching etc. that we need to pay special attention to

kimishpatel replied:
See my response above. For dynamic quantization, you probably can but for static you cant just follow those rules.

jerryzh168 reviewed (changes requested):
thanks, looks good in the high level, I feel we should be able to remove the dependency on annotation and qspec, also I think that could potentially simplify the implementation a bit, let me know what do you think

kimishpatel commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,215 @@
+    return None
+
+
+def _port_metadata_for_input_quant_nodes(
```
Also what if you want to dynamically quantize output. If you were to tag that information without looking at the qspec, you might make incorrect associations

jerryzh168 commented on test/quantization/pt2e/test_metadata_porting.py:
```diff
@@ -0,0 +1,384 @@
+                pass
+
+        example_inputs = (torch.randn(1, 3, 5, 5),)
+        quantizer_per_tensor_tags = set(
```
nit: quantize_per_tensor_tags? same for `dequantizer_per_tensor_tags` and `dequantizer_per_channel_tags`

jerryzh168 commented on test/quantization/pt2e/test_metadata_porting.py:
```diff
@@ -0,0 +1,384 @@
+        """
+        Model under test
+        conv2d -> avgpool -> hardtanh -> linear
+        Quantize all except linear.
```
nit: Quantize all with static quant except linear?

jerryzh168 reviewed (approved):
Record some of the things we were discussing here so that we don't forget:

1. right now port meta pass is relying on "quantization_annotation" right now, although I think the ideally we can do: https://github.com/pytorch/pytorch/pull/107107/files#r1295252189, that will make the information flow cleaner and easier for debugging as well, it will also prevent future uses quantization_annotation after prepare step (we could also just delete quantization_annotation in the future to prevent this), to preserve the information needed by https://github.com/pytorch/pytorch/pull/107107/files#r1295244618, we could record something like add_node.meta["precision"] = torch.float32 so that no metadata porting happens for add node. It's not clear from Edge side if this is strictly needed, also this is not blocking anything right now, so we can revisit this when this actually matters for some use cases.

2. right now the traversal to find the "choose_qparams -> q -> dq" and "q -> dq" patterns is from producer to consumer, but I feel the reversed order might be more robust, we can first start with just looking at the graph (i.e. identify choose_qparams -> q -> dq and q -> dq patterns) and then look around the operators that appears before and after the pattern + rely on some of the quantization_annotation information to decide whether to port metadata or not. I think this will be aligned better with the node.meta["precision"] solution, and the use of quantization annotation will be relatively minimized as well

We have agreed that user should not assume "quantization_annotation" being available in convert step. also given that we might need to align more on (1)., I'll just stamp for now and we can revisit the implementation details at a later time when this comes up. For now @kimishpatel / Edge team will maintain this code util we have reached an agreement on (1).

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,215 @@
+    - quantizer_per_tensor.tensor, dequantize_per_tensor.tensor
+    - quantizer_per_channel.default, dequantize_per_channel.default
+
+    Rules of porting metadata:
```
please make sure to update the docs to mention that we are relying on quantization_annotation right now, and document that we are not settled on quantization_annotation being available at this step as well

kimishpatel commented:
> user should not assume "quantization_annotation" being available in convert step

while this is fair, note that the metadata porting is part of the convert step, so this is not strictly user facing.

So if "quantization_annotation" is something that has to be become user visible, in order to convey whether a set of nodes were meant to be quantized or not, then I agree that it may not be the right choice and maybe something like what you suggested, node.meta["precision"] or equivalent, might be a better way to do it.

But if it is mostly internal to quantization workflow, then I am not certain of the value added by node.meta["precision"], because that will be derived from quantization_annotation anyway.

jerryzh168 commented:
> > user should not assume "quantization_annotation" being available in convert step
> 
> while this is fair, note that the metadata porting is part of the convert step, so this is not strictly user facing.
> 
> So if "quantization_annotation" is something that has to be become user visible, in order to convey whether a set of nodes were meant to be quantized or not, then I agree that it may not be the right choice and maybe something like what you suggested, node.meta["precision"] or equivalent, might be a better way to do it.
> 
> But if it is mostly internal to quantization workflow, then I am not certain of the value added by node.meta["precision"], because that will be derived from quantization_annotation anyway.

yeah I'm not sure if the node.meta["precision"] info will be needed by users or not, so it can be a discussion left for future discussions. Apart from this, I think the benefit is mostly BE, easier to debug and reduce the chances that people rely on quantization_annotation, as specified in the first point.

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -0,0 +1,227 @@
+
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        for node in graph_module.graph.nodes:
+            annotation = node.meta.get("quantization_annotation", None)
```
@kimishpatel could you also add comment here saying that even we are relying on quantization annotation here, but backend developers should not assume "quantization_annotation" being available after convert?

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).

kimishpatel committed:
Update on "[Quantization] Add metadata porting for nodes added by quantization"


Summary:
This diff adds adding metadata to q-dq nodes by inferring the
quatization intent from node annotations. Annotations on the node are
way for user to specify how a node or subgraph is supposed to be
quantized. We continue to use that information to copy metadata on Q/DQ
node from appropriate nodes.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48488416](https://our.internmc.facebook.com/intern/diff/D48488416)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48488416).
